chain/client/src/client.rs

```diff
@@ -2140,3 +2152,80 @@ impl Client {
         }
     }
 }
+
+impl Client {
+    /// Each epoch defines a set of important accounts: block producers, chunk producers,
+    /// approvers. Low-latency reliable communication between those accounts is critical,
+    /// so that the blocks can be produced on time. This function computes the set of
+    /// important accounts (aka TIER1 accounts) so that it can be fed to PeerManager, which
+    /// will take care of the traffic prioritization.
+    ///
+    /// It returns both TIER1 accounts for both current epoch (according to the `tip`)
+    /// and the next epoch, so that the PeerManager can establish the priority connections
+    /// in advance (before the epoch starts and they are actually needed).
+    ///
+    /// The result of the last call to get_tier1_accounts() is cached, so that it is not recomputed
+    /// if the current epoch didn't change since the last call. In particular SetChainInfo is send
+    /// after every block is processed (order of seconds), while the epoch changes way less
+    /// frequently (order of hours).
+    fn get_tier1_accounts(&mut self, tip: &Tip) -> Result<Arc<AccountKeys>, Error> {
```

Comment:
As you say, the current implementation is flexible enough to account for removing slashed accounts within an epoch. Also, having AccountKeys indexed by epoch_id would make it easy to make this bug, just as you pointed out.

Reply:
Sending a message with optional field to indicate "nothing" changed, is isomorphic to having 2 separate message types. However it obfuscates the difference between "not sending an update" and "sending an empty update", so I'd have rather 2 separate message types if we went that way.

Reply:
>  and later at the PeerManagerActor side, it needs to check if the field changed

PeerManager will check if anything changed anyway, because a change triggers a full AccountsData sync with all the connected peers (which is expensive) and I don't want Client to be responsible for how efficiently the PeerManager behaves.

Reply:
> and also, then the cache here wouldn't work either.

true

Reply:
I can't say that the current implementation feels like the obviously best one, but I also don't see anything obviously better. 

What we want to do here is essentially synchronize some state between two components running concurrently. I would say there are roughly three different patterns how to achieve this: 

* fine-grained callbacks/observers, where a specific callback is invoked when a specific bit of state changes 
* snapshot streaming, where reify the whole shared state into a single datastructure, and periodically send this whole state as a single snapshot
* optimized snapshot streaming, where the API looks as if the source pushed a snapshot, and the destination pulls a snapshot, but where on the transport layer we actually send the diff with the most recent snapshot

The main tradeoff here is runtime efficiency vs programming simplicity: callbacks are more efficient, because they tell you exactly what changed. Snapshots are easier because:

* on the sending side, you don't have to be super-precise on when exactly are you sending a new snapshot, as long as you eventually do that
* on the receiving side, you get a weaker programming model: you only know the current state, not the sequence of sets which led there. This forces simpler code which can't be broken by the changes on the sender side
* there's also a specific `struct Snapshot` which you can point your finger at and say "that's the whole state that's shared"


Snapshots are also easier to do if the communication channel is unreliable, but I think this is not applicable to us: `do_send` does guarantee message delivery within a single process, actix is not *that* crazy :) 

I do find it odd that we do the "did this change?" check twice: on the sender side, when we look in the cache, and on the receiver side, when we use `Arc::eq`. It does seem more robust to include some sort of the version field there. But I don't think this is going to be a *huge* problem: this only affects efficiency, not correctness. Ultimately, I hope we do get rid of actix soonish, and, once we do that, this pattern could perhaps be made more explicit with the help of  https://docs.rs/tokio/1.20.1/tokio/sync/watch/index.html.


It also admittedly is buried in details that we do `Arc::ptr_eq` here. It's obvious in the diff, but won't be so obvious if you reed the code. 

One thing we could do is to move diffing to the `ChainInfo` struct, where it belongs: 

```diff
diff --git a/chain/network-primitives/src/types.rs b/chain/network-primitives/src/types.rs
index 120d8217c..f063ca9a0 100644
--- a/chain/network-primitives/src/types.rs
+++ b/chain/network-primitives/src/types.rs
@@ -341,6 +341,12 @@ pub struct ChainInfo {
     pub tier1_accounts: Arc<AccountKeys>,
 }
 
+impl ChainInfo {
+    pub fn are_tier1_accounts_updated(&self, old: &ChainInfo) -> bool {
+        !Arc::ptr_eq(&self.tier1_accounts, &old.tier1_accounts)
+    }
+}
+
 #[derive(Debug, actix::Message)]
 #[rtype(result = "()")]
 pub struct SetChainInfo(pub ChainInfo);
diff --git a/chain/network/src/accounts_data/mod.rs b/chain/network/src/accounts_data/mod.rs
index defe3cc82..7a1385b98 100644
--- a/chain/network/src/accounts_data/mod.rs
+++ b/chain/network/src/accounts_data/mod.rs
@@ -150,15 +150,10 @@ impl Cache {
     /// Updates the set of important accounts and their public keys.
     /// The AccountData which is no longer important is dropped.
     /// Returns true iff the set of accounts actually changed.
-    pub fn set_keys(&self, keys: Arc<AccountKeys>) -> bool {
+    pub fn set_keys(&self, keys: Arc<AccountKeys>) {
         let mut inner = self.inner.write();
-        // Skip further processing if the key set didn't change.
-        if Arc::ptr_eq(&keys, &inner.keys) || keys == inner.keys {
-            return false;
-        }
         inner.data.retain(|k, _| keys.contains_key(k));
         inner.keys = keys;
-        true
     }
 
     /// Selects new data and verifies the signatures.
diff --git a/chain/network/src/peer_manager/peer_manager_actor.rs b/chain/network/src/peer_manager/peer_manager_actor.rs
index f2d29c960..2b45e8f06 100644
--- a/chain/network/src/peer_manager/peer_manager_actor.rs
+++ b/chain/network/src/peer_manager/peer_manager_actor.rs
@@ -2129,12 +2129,21 @@ impl PeerManagerActor {
 
 impl Handler<SetChainInfo> for PeerManagerActor {
     type Result = ();
-    fn handle(&mut self, info: SetChainInfo, _ctx: &mut Self::Context) {
-        if self.state.accounts_data.set_keys(info.0.tier1_accounts.clone()) {
+    fn handle(&mut self, SetChainInfo(new_info): SetChainInfo, _ctx: &mut Self::Context) {
+        let mut new_accounts = None;
+        {
+            let guard = self.state.chain_info.write();
+            if new_info.are_tier1_accounts_updated(&guard) {
+                new_accounts = Some(Arc::clone(&new_info.tier1_accounts));
+            }
+            *guard = new_info
+        }
+
+        if let Some(new_accounts) = new_accounts {
+            self.state.accounts_data.set_keys(new_accounts);
             // TODO(gprusak): the set of tier1 accounts has changed.
             // We might miss some data, so start a full sync with connected peers.
         }
-        *self.state.chain_info.write() = info.0;
     }
 }
 
```